## Exploratory Data Analysis
* Checking Data
* Statistical Summary
* Univariate Analysis
* Bivariate and Multivariate Analysis
* Deep Dive Exploration
* EDA Conclusion

In [2]:
import pandas as pd
import numpy as np

# Checking Data

In [3]:
df = pd.read_csv("/Users/darshmac/Desktop/Telecom_Dataset.csv")

In [4]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [5]:
df.sample(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
3175  7395-IGJOS    Male              1     Yes         No      17   
6598  6169-PGNCD  Female              0      No         No      57   
3194  6397-JNZZG  Female              1     Yes         No      43   
1369  5478-JJVZK  Female              0      No         No      33   
692   9620-QJREV    Male              0      No        Yes      47   
3069  1784-BXEFA  Female              0      No         No      10   
6265  3278-FSIXX  Female              0      No         No       2   
5930  8398-TBIYD  Female              0      No         No      72   
6278  7208-PSIHR  Female              0     Yes         No      70   
2271  0093-EXYQL  Female              1      No         No      40   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
3175          Yes                No     Fiber optic                   No  ...   
6598          Yes                No             DSL                  Yes  ...   
3194           No  No phone service             DSL                   No  ...   
1369          Yes               Yes             DSL                  Yes  ...   
692           Yes                No              No  No internet service  ...   
3069          Yes                No              No  No internet service  ...   
6265          Yes                No             DSL                   No  ...   
5930          Yes               Yes     Fiber optic                  Yes  ...   
6278          Yes               Yes     Fiber optic                   No  ...   
2271          Yes                No     Fiber optic                   No  ...   

         DeviceProtection          TechSupport          StreamingTV  \
3175                   No                   No                   No   
6598                  Yes                  Yes                  Yes   
3194                  Yes                   No                  Yes   
1369                   No                   No                   No   
692   No internet service  No internet service  No internet service   
3069  No internet service  No internet service  No internet service   
6265                   No                   No                   No   
5930                   No                  Yes                   No   
6278                  Yes                  Yes                  Yes   
2271                   No                   No                  Yes   

          StreamingMovies        Contract PaperlessBilling  \
3175                  Yes  Month-to-month               No   
6598                   No        Two year              Yes   
3194                  Yes  Month-to-month               No   
1369                   No  Month-to-month              Yes   
692   No internet service        One year               No   
3069  No internet service  Month-to-month              Yes   
6265                   No  Month-to-month              Yes   
5930                   No        Two year              Yes   
6278                  Yes        Two year              Yes   
2271                  Yes  Month-to-month              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
3175  Bank transfer (automatic)          81.50        1329.2   Yes  
6598    Credit card (automatic)          74.30       4166.35    No  
3194    Credit card (automatic)          55.55        2342.2   Yes  
1369               Mailed check          60.90       2033.85    No  
692     Credit card (automatic)          20.55       1013.05    No  
3069               Mailed check          20.55        187.35    No  
6265               Mailed check          45.25          74.2    No  
5930  Bank transfer (automatic)          84.70       6185.15    No  
6278  Bank transfer (automatic)         104.30        7188.5    No  
2271           Electronic check          91.55        3673.6    No  

[10 rows x 21 columns]

Let's see unique value of some feature

In [25]:
df.shape

(7032, 21)

In [7]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In the TotalCharges feature you can see the data type is Object, it should be numeric. So the data type must be changed.

In [13]:
df = df.loc[~df['TotalCharges'].str.contains(' ')]

In [14]:
df['TotalCharges'] = df['TotalCharges'].astype(float)

/var/folders/yp/tcxqwqns2w52l01lkcdwsy6w0000gn/T/ipykernel_2078/3174716150.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['TotalCharges'] = df['TotalCharges'].astype(float)


In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 


The Data Type of the TotalCharges feature has changed to float

## Check missing value and duplicated data

In [16]:
df.isna().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [17]:
df.duplicated().sum()

0

## Observation on Checking Data
* The data consists of 21 columns and 7043 rows
* There's no missing value and duplicated data
* The data type in the Total Charge feature from string is changed to float
* There are 4 columns which are numerical, SeniorCitizen, tenure, MonthlyCharges, TotalCharges. Apart from the 4 columns, there are categories.

# Statistical Summary

In [20]:
numeric_columns = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_columns = df.select_dtypes(include=['object']).columns.tolist()

print("Numeric Columns:", numeric_columns)
print("Object Columns:", categorical_columns)

Numeric Columns: ['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges']
Object Columns: ['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn']


In [21]:
df[numeric_columns].describe()

SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7032.000000  7032.000000     7032.000000   7032.000000
mean        0.162400    32.421786       64.798208   2283.300441
std         0.368844    24.545260       30.085974   2266.771362
min         0.000000     1.000000       18.250000     18.800000
25%         0.000000     9.000000       35.587500    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.862500   3794.737500
max         1.000000    72.000000      118.750000   8684.800000

## Observation of Numericals
* The Min and Max values ​​for each column still make sense
* The Senior Citizen column only has values 0 and 1

In [22]:
df[categorical_columns].describe()

customerID gender Partner Dependents PhoneService MultipleLines  \
count         7032   7032    7032       7032         7032          7032   
unique        7032      2       2          2            2             3   
top     7590-VHVEG   Male      No         No          Yes            No   
freq             1   3549    3639       4933         6352          3385   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7032           7032         7032             7032   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3497         3087             3094   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7032        7032            7032            7032   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3472        2809            2781            3875   

       PaperlessBilling     PaymentMethod Churn  
count              7032              7032  7032  
unique                2                 4     2  
top                 Yes  Electronic check    No  
freq               4168              2365  5163

## Observation of Categoricals
* Customers are dominated by male gender
* Columns Partner, Dependents, PhoneService, and MultipleLines only have 2 values, Yes and No.
* Most Internet Services use Fiber Optic
* The contract most used by customers is Month-to-month
* The PaymentMethod most widely used by customers is Electronic check
* Many customers have No Churn

In [23]:
df['PaymentMethod'].value_counts()

Electronic check             2365
Mailed check                 1604
Bank transfer (automatic)    1542
Credit card (automatic)      1521
Name: PaymentMethod, dtype: int64